src/differentiation/compute_jacobian_ad.jl

```diff
@@ -94,16 +94,18 @@ end
 end
 
 function forwarddiff_color_jacobian(f,x::AbstractArray{<:Number},jac_cache::ForwardColorJacCache,jac_prototype=nothing)
-    dx = jac_cache.dx
-    vecx = vec(x)
-    sparsity = jac_cache.sparsity
 
-    J = jac_prototype isa Nothing ? (sparsity isa Nothing ? false .* vec(dx) .* vecx' : zeros(eltype(x),size(sparsity))) : zero(jac_prototype)
+    if jac_prototype isa Nothing ? ArrayInterface.ismutable(x) : ArrayInterface.ismutable(jac_prototype)
+        # Whenever J is mutable, we mutate it to avoid allocations
+        dx = jac_cache.dx
+        vecx = vec(x)
+        sparsity = jac_cache.sparsity
 
-    if ArrayInterface.ismutable(J) # Whenever J is mutable, we mutate it to avoid allocations
+        J = jac_prototype isa Nothing ? (sparsity isa Nothing ? false .* vec(dx) .* vecx' :
+                                         zeros(eltype(x),size(sparsity))) : zero(jac_prototype)
         forwarddiff_color_jacobian(J, f, x, jac_cache)

```

Comment:
this branch needs to `return J` afterwards as well, otherwise it's relying on the last line of `forwarddiff_color_jacobian` to return `J` which isn't necessarily guaranteed from a mutating function.